### PR TITLE
Hyper search state

### LINKED ISSUE #2494: Implement HyperSearchState Interface
Implement a module for storing the state of hypersearch runs (abstract class to be subclassed)

### PR DESCRIPTION
**Pull request by mihail911**
Skeleton for HypersearchState. @scottpurdy 
Fixes #2494

### CODE REVIEW

**Comment by mihail911** [comment]
Fixes @2494

File: nupic/swarming/hypersearch/HyperSearchState.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+                                      evaluating models/running jobs.
+    """
+    # Number of particles in current running instance of swarm
+    self.numParticles = numParticles
+    # Store collection of user-specified permutation variables
+    self.permuteVars = permuteVars
+    self.swarmIDs = []
+    # Store collection of particles for given swarm, where key is
+    # the swarmID and value is the list of particles for the swarm
+    self.particles = dict()
```
Comment (scottpurdy) [inline nupic/swarming/hypersearch/HyperSearchState.py]: Use `{}` just like you do for list literals. This avoids a "list" lookup.

File: nupic/swarming/hypersearch/HyperSearchState.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    Defines the class storing necessary state for hyperSearch runs.
+  """
+
+  def __init__(self, permuteVars, particles, numParticles, options):
+    """
+      @oaram  permuteVars   (list)    Collection of client-specified permutation
+                                      variables necessary for a given run.
+      @param  particles     (list)    Collection of particles instantiated by
+                                      client, in HyperSearchRunner.
+      @param  numParticles  (int)     Client provided number of particles in
```
Comment (scottpurdy) [inline nupic/swarming/hypersearch/HyperSearchState.py]: Doesn't this duplicate `len(particles)`

**Comment by scottpurdy** [comment]
@oxtopus - please review

File: nupic/swarming/hypersearch/HyperSearchState.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+                                      swarm. TODO: May want to just provide a
+                                      generic param with swarm-specific options?
+      @param  options       (dict)    Dict of client-specified params for
```
Comment (oxtopus) [inline nupic/swarming/hypersearch/HyperSearchState.py]: This should either be broken up into separate kwargs, or be an instance of a specific class representing possible parameters.

File: nupic/swarming/hypersearch/HyperSearchState.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    # Number of particles in current running instance of swarm
+    self.numParticles = numParticles
+    # Store collection of user-specified permutation variables
+    self.permuteVars = permuteVars
+    self.swarmIDs = []
+    # Store collection of particles for given swarm, where key is
+    # the swarmID and value is the list of particles for the swarm
+    self.particles = dict()
+    # Dict of params used to generate particular model
+    self.options= options
```
Comment (oxtopus) [inline nupic/swarming/hypersearch/HyperSearchState.py]: Spacing.

File: nupic/swarming/hypersearch/HyperSearchState.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    self.modelIDs = []
+
+    # TODO: Define custom JobInfo module with info about all running jobs?
+    # This instance variable would then be a dict of (jobID, JobInfo) instances
+    self.jobStates = dict()
+
+    # TODO: Ditto
+    self.modelStates = dict()
+
+  def getParticleStates(self):
```
Comment (oxtopus) [inline nupic/swarming/hypersearch/HyperSearchState.py]: Where is this function used?  I suspect it isn't needed.  If it is, I would consider making it a generator:

```python
for particle in self.particles:
  yield particle.getState()
```

**Comment by scottpurdy** [comment]
We may need to hold off on this until we have the formal design doc.

File: nupic/swarming/hypersearch/HyperSearchState.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+
+  def updateState(self, swarmID, score):
+    """
+      Update score for given swarm.
+
+      @param  swarmID   (int)     ID of swarm to be updated
+      @param  score     (int)     Score to update for swarm
+    """
+    self.bestScore[swarmID] = score
```
Comment (oxtopus) [inline nupic/swarming/hypersearch/HyperSearchState.py]: Should this not be conditional?  i.e. self.bestScore[swarmID] = max(score, 

```python
currentBest = self.bestScore[swarmID]
if score > currentBest:
  self.bestScore[swarmID] = score
```

Or:

```python
self.bestScore[swarmID] = max(self.bestScore[swarmID], score)
```

**Pull request closed without merging**